README
Sample readme file for Scratch3-splash project.


NEWS
Sample NEWS file for Scratch3-splash project.


cleanup.sh
#!/bin/sh -e

# Source and credits:
# https://github.com/dblommesteijn/autotools-example

# collect input
while true; do
	read -p "Do you wish to remove all maintainer files [y/n]? " yn
	case $yn in
		[Yy]* ) break;;
		[Nn]* ) exit;;
		* ) echo "Please answer y or n.";;
	esac
done

# dirs
rm -rf autom4te.cache/
rm -rf aux-dist/
rm -rf m4/
rm -rf src/.deps
rm -rf src/.libs
rm -f src/*.lo
rm -f src/*.la
rm -rf src/libtool
rm -rf src/toollib/.deps
rm -f src/toollib/.dirstamp
rm -f src/toollib/*.lo
rm -f src/toollib/*.la
# files
if [ -f include/Makefile ]; then
	rm include/Makefile
fi
if [ -f include/Makefile.in ]; then
	rm include/Makefile.in
fi
if [ -f src/Makefile ]; then
	rm src/Makefile
fi
if [ -f src/Makefile.in ]; then
	rm src/Makefile.in
fi
if [ -f aclocal.m4 ]; then
	rm aclocal.m4
fi
if [ -f config.status ]; then
	rm config.status
fi
if [ -f config.log ]; then
	rm config.log
fi
if [ -f configure ]; then
	rm configure
fi
if [ -f libtool ]; then
	rm libtool
fi
if [ -f Makefile ]; then
	rm Makefile
fi
if [ -f Makefile.in ]; then
	rm Makefile.in
fi
if [ -f config.guess ]; then
        rm config.guess
fi
if [ -f config.sub ]; then
        rm config.sub
fi
if [ -f depcomp ]; then
        rm depcomp
fi
if [ -f install-sh ]; then
        rm install-sh
fi
if [ -f missing ]; then
        rm missing
fi


echo "Cleaned files!"


src/TcpClient.h
/*
 * TcpClient.h
 *
 *  Created on: Mar 2, 2019
 *      Author: peter
 */

#ifndef TCPCLIENT_H_
#define TCPCLIENT_H_

#include <iostream>
#include <istream>
#include <ostream>
#include <string>
#include <boost/asio.hpp>
#include <boost/bind.hpp>

using boost::asio::ip::tcp;

class TcpClient
{
public:
	TcpClient(boost::asio::io_service& io_service, const std::string& server, const std::string& path);
	virtual ~TcpClient();
	
private:
	void handle_resolve(const boost::system::error_code& err, tcp::resolver::iterator endpoint_iterator);
	void handle_connect(const boost::system::error_code& err);
	void handle_write_request(const boost::system::error_code& err);
	void handle_read_status_line(const boost::system::error_code& err);
	void handle_read_headers(const boost::system::error_code& err);
	void handle_read_content(const boost::system::error_code& err);
	
	tcp::resolver resolver_;
	tcp::socket socket_;
	boost::asio::streambuf request_;
	boost::asio::streambuf response_;
};

#endif /* TCPCLIENT_H_ */


src/MainWindow.cpp
/*
 * MainWindow.cpp
 *
 *  Created on: Mar 1, 2019
 *      Author: peter
 */

#include "MainWindow.h"
#include <gtkmm-3.0/gtkmm/window.h>

void thread(MainWindow *mw)
{
	
	sleep(3);
	
	boost::asio::io_context::count_type count = -2;
	for (int i = 0; i < 10; ++i) 
	{
		try
		{
			boost::asio::io_service io_service;
			TcpClient c(io_service, "localhost", "/");
			count = io_service.run();
		} catch (std::exception& e)
		{
			std::cout << "Exception: " << e.what() << "\n";
			count = -1;
		}
		if (count >= 6 || count == -1)
			break;
		if (!mw)
		{
			std::cout << "Error: Terminating." << count << std::endl;
			exit(1);
		}
		sleep(1);
	}
	
	std::cout << "Done1!" << count << std::endl;
	mw->close();
}

MainWindow::MainWindow() :
		m_splashImage(std::string(IMAGE_DATA_DIR) + "/scratch.jpg")
//	m_addrInfo(NULL),
//	m_sockFD(0)
//	m_curl(curl_easy_init())
{
	std::cout << std::string(IMAGE_DATA_DIR) + "/scratch.jpg" << std::endl;
	set_size_request(800, 600);
	add(m_splashImage);
	//fullscreen();
	set_keep_above(true);
	m_splashImage.show();
	
	//signalConnected().connect()

	//struct curl_slist *connect_to = NULL;
	//connect_to = curl_slist_append(NULL, "example.com::server1.example.com:");

	//curl_easy_setopt(m_curl, CURLOPT_URL, "http://localhost:8601");
	//curl_easy_perform(m_curl);
	//curl_multi_timeout

	/*addrinfo hints;
	 memset(&hints, 0, sizeof hints);
	 hints.ai_family = AF_UNSPEC;
	 hints.ai_socktype = SOCK_STREAM;*/
	
	std::cout << "Done2!" << std::endl;
	
}

MainWindow::~MainWindow()
{
	std::cout << "~MainWindow()" << std::endl;
}

/*MainWindow::siganelConnected_t MainWindow::signalConnected(void)
{
	return m_siganelConnected;
}*/


src/tmp/borderless.cc
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<X11/X.h>
#include<X11/Xlib.h>
#include<GL/gl.h>
#include<GL/glx.h>
#include<GL/glu.h>
 
Display                 *dpy;
Window                  root, win;
GLint                   att[] = { GLX_RGBA, GLX_DEPTH_SIZE, 24, GLX_DOUBLEBUFFER, None };
Colormap                cmap;
XVisualInfo             *vi;
GLXContext              glc;
 
int main(int argc, char *argv[]) {
 XEvent                 xev;
 XWindowAttributes      gwa;
 XSetWindowAttributes   swa;
 int                    cw_mask = CWEventMask | CWColormap | CWOverrideRedirect;
 
 dpy    = XOpenDisplay(NULL);
 root   = DefaultRootWindow(dpy);
 vi     = glXChooseVisual(dpy, 0, att);
 cmap   = XCreateColormap(dpy, root, vi->visual, AllocNone);
 glc    = glXCreateContext(dpy, vi, NULL, GL_TRUE);
 
 swa.override_redirect = True;
 swa.event_mask = ExposureMask;
 swa.colormap = cmap;
 
 win = XCreateWindow(dpy, root, 200, 200, 400, 400, 0, vi->depth, InputOutput, vi->visual, cw_mask, &swa);
 
 XMapWindow(dpy, win);
 glXMakeCurrent(dpy, win, glc);
 
 while(1) {
        XNextEvent(dpy, &xev);
 
        if(xev.type == Expose) {
                XGetWindowAttributes(dpy, win, &gwa);
                glViewport(0, 0, gwa.width, gwa.height);
                glClearColor(0.3, 1.0, 0.3, 1.0);
                glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
 
                glXSwapBuffers(dpy, win); } } }
//
//      gcc -o borderless borderless.cc -lX11 -lGL -lGLU
//


src/TcpClient.cpp
/*
 * TcpClient.cpp
 *
 *  Created on: Mar 2, 2019
 *      Author: peter
 */

#include "TcpClient.h"

TcpClient::TcpClient(boost::asio::io_service& io_service, const std::string& server, const std::string& path) :
		resolver_(io_service), socket_(io_service)
{
	// Form the request. We specify the "Connection: close" header so that the
	// server will close the socket after transmitting the response. This will
	// allow us to treat all data up until the EOF as the content.
	std::ostream request_stream(&request_);
	request_stream << "GET " << path << " HTTP/1.0\r\n";
	request_stream << "Host: " << server << "\r\n";
	request_stream << "Accept: */*\r\n";
	request_stream << "Connection: close\r\n\r\n";

	// Start an asynchronous resolve to translate the server and service names
	// into a list of endpoints.
	tcp::resolver::query query(server, "8601");
	resolver_.async_resolve(query, boost::bind(&TcpClient::handle_resolve, this, boost::asio::placeholders::error, boost::asio::placeholders::iterator));
}

TcpClient::~TcpClient()
{
}

void TcpClient::handle_resolve(const boost::system::error_code& err, tcp::resolver::iterator endpoint_iterator)
{
	std::cout << "handle_resolve" << std::endl;
	if (!err)
	{
		// Attempt a connection to each endpoint in the list until we
		// successfully establish a connection.
		boost::asio::async_connect(socket_, endpoint_iterator, boost::bind(&TcpClient::handle_connect, this, boost::asio::placeholders::error));
	}
	else
	{
		std::cout << "Error: " << err.message() << "\n";
	}
}

void TcpClient::handle_connect(const boost::system::error_code& err)
{
	std::cout << "handle_connect" << std::endl;
	if (!err)
	{
		// The connection was successful. Send the request.
		boost::asio::async_write(socket_, request_, boost::bind(&TcpClient::handle_write_request, this, boost::asio::placeholders::error));
	}
	else
	{
		std::cout << "Error: " << err.message() << "\n";
	}
}

void TcpClient::handle_write_request(const boost::system::error_code& err)
{
	std::cout << "handle_write_request" << std::endl;
	if (!err)
	{
		// Read the response status line. The response_ streambuf will
		// automatically grow to accommodate the entire line. The growth may be
		// limited by passing a maximum size to the streambuf constructor.
		boost::asio::async_read_until(socket_, response_, "\r\n", boost::bind(&TcpClient::handle_read_status_line, this, boost::asio::placeholders::error));
	}
	else
	{
		std::cout << "Error: " << err.message() << "\n";
	}
}

void TcpClient::handle_read_status_line(const boost::system::error_code& err)
{
	std::cout << "handle_read_status_line" << std::endl;
	if (!err)
	{
		// Check that response is OK.
		std::istream response_stream(&response_);
		std::string http_version;
		response_stream >> http_version;
		unsigned int status_code;
		response_stream >> status_code;
		std::string status_message;
		std::getline(response_stream, status_message);
		if (!response_stream || http_version.substr(0, 5) != "HTTP/")
		{
			std::cout << "Invalid response\n";
			return;
		}
		if (status_code != 200)
		{
			std::cout << "Response returned with status code ";
			std::cout << status_code << "\n";
			return;
		}

		// Read the response headers, which are terminated by a blank line.
		boost::asio::async_read_until(socket_, response_, "\r\n\r\n", boost::bind(&TcpClient::handle_read_headers, this, boost::asio::placeholders::error));
	}
	else
	{
		std::cout << "Error: " << err << "\n";
	}
}

void TcpClient::handle_read_headers(const boost::system::error_code& err)
{
	std::cout << "handle_read_headers" << std::endl;
	if (!err)
	{
		// Process the response headers.
		std::istream response_stream(&response_);
		std::string header;
		while (std::getline(response_stream, header) && header != "\r")
			std::cout << header << "\n";
		std::cout << "\n";

		// Write whatever content we already have to output.
		if (response_.size() > 0)
			std::cout << &response_;

		// Start reading remaining data until EOF.
		boost::asio::async_read(socket_, response_, boost::asio::transfer_at_least(1), boost::bind(&TcpClient::handle_read_content, this, boost::asio::placeholders::error));
	}
	else
	{
		std::cout << "Error: " << err << "\n";
	}
}

void TcpClient::handle_read_content(const boost::system::error_code& err)
{
	std::cout << "handle_read_content" << std::endl;
	if (!err)
	{
		// Write all of the data that has been read so far.
		std::cout << &response_;

		// Continue reading remaining data until EOF.
		boost::asio::async_read(socket_, response_, boost::asio::transfer_at_least(1), boost::bind(&TcpClient::handle_read_content, this, boost::asio::placeholders::error));
	}
	else if (err != boost::asio::error::eof)
	{
		std::cout << "Error: " << err << "\n";
	}
}
